keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ResourceNotifier.kt

```diff
@@ -0,0 +1,149 @@
+
+  private val notificationsEnabled: Boolean
+    get() = springEnv.getProperty("keel.notifications.resource", Boolean::class.java, true)
+
+  @EventListener(ResourceNotificationEvent::class)
+  fun onResourceNotificationEvent(event: ResourceNotificationEvent) {
+    if (notificationsEnabled){
+      val shouldNotify = notifierRepository.addNotification(event.scope, event.identifier, event.notifier)
+      if (shouldNotify) {
+        notify(event)
+        notifierRepository.markSent(event.scope, event.identifier, event.notifier)
+        spectator.counter(
+          NOTIFICATION_SENT_ID,
+          listOf(BasicTag("notifier", event.notifier.name))
+        )
+          .runCatching { increment() }
+          .onFailure {
+            log.error("Exception incrementing {} counter: {}", NOTIFICATION_SENT_ID, it.message)
+          }
+      }
+    }
+  }
+
+  /**
+   * This method assumes we're notifying about a resource.
+   * If / when we notify about something else, we will need to update this assumption
+   */
+  private fun notify(event: ResourceNotificationEvent) {
+    log.debug("Sending notifications for ${event.scope.name.toLowerCase()} ${event.identifier} with content ${event.message}")
+    try {
+      val application = keelRepository.getResource(event.identifier).application
+      val env = keelRepository.environmentFor(event.identifier)
+      env.notifications.forEach { notificationConfig ->
+        val notification = notificationConfig.toEchoNotification(application, event)
+        log.debug("Sending notification for resource ${event.identifier} with config $notification")
+        runBlocking {
+          echoService.sendNotification(notification)
+        }
+      }
+    } catch (e: Exception) {
+      when (e) {
+        is NoSuchResourceException -> {
+          log.error("Trying to notify resource ${event.identifier} but it doesn't exist. Not notifying.")
+          return
+        }
+        is OrphanedResourceException -> {
+          log.error("Trying to notify resource ${event.identifier} but it doesn't have an environment. Not notifying.")
+          return
+        }
+        else -> throw e
+      }
+    }
+
+  }
+
+  private fun NotificationConfig.toEchoNotification(application: String, event: ResourceNotificationEvent): EchoNotification {
+    return EchoNotification(
+      notificationType = EchoNotification.Type.valueOf(type.name.toUpperCase()),
+      to = listOf(address),
+      severity = EchoNotification.Severity.NORMAL,
+      source = EchoNotification.Source(
+        application = application
+      ),
+      additionalContext = mapOf(
+        "formatter" to "MARKDOWN",
+        "subject" to event.message.subject,
+        "body" to event.message.body
+      ),
+      interactiveActions = generateInteractiveConfig(event.identifier, event.notifier.name, event.message.color)
+    )
+  }
+
+  private fun generateInteractiveConfig(resourceId: String, notifier: String, color: String): InteractiveActions? =
```

Comment:
We might want to fix the model in echo such that you don't have to specify a whole `interactiveActions` block just to specify the color, but that's icing on the cake. 🙂

Reply:
Sure, we could do that in a follow up PR. It doesn't feel that important though